# PROJECT TITLE : Data Science Jobs Analysis on Glassdoor (USA)

# End Goals of Project
1. Determining the mean salaries of data-related positions throughout the United States.

2. Identifying the prevalent technologies in data jobs.

3. Determining the most commonly used Business Intelligence (BI) tool.

4. Calculating the percentage of top data jobs.

5. Identifying the sectors with the highest employment of data-related jobs.

### Target audience
Data professionals and job seekers: Individuals working in or seeking data-related positions who want to understand the salary trends, prevalent technologies, demand for specific data jobs, and popular sectors employing data talent.




# Procedures to be taken
* Data Cleaning
* Data transformation (some will be done in Power BI)
* Data Visualization

### Importing necessary Libraries

In [100]:
# Import necessary libraries
import missingno as msno
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import seaborn as sns
import re

### Loading the csv into Jupyter Lab

In [101]:
# Reading csv
python_file = pd.read_csv('Uncleaned_DS_jobs.csv', index_col=0)

### Converting to DataFrame using Pandas pd.Dataframe

In [102]:
# Convert to dataframe for easy cleaning 
csv_dataframe=pd.DataFrame(python_file)

### Overview of Dataset to Diagnose problem of Dirty Data

In [103]:
# Display the first 50 rows of the dataframe to diagnose problem of dirty data
csv_dataframe.head(50)

Job Title  \
index                                                   
0                                   Sr Data Scientist   
1                                      Data Scientist   
2                                      Data Scientist   
3                                      Data Scientist   
4                                      Data Scientist   
5                                      Data Scientist   
6            Data Scientist / Machine Learning Expert   
7                                      Data Scientist   
8                    Staff Data Scientist - Analytics   
9                                      Data Scientist   
10                                     Data Scientist   
11                                     Data Scientist   
12          Data Scientist - Statistics, Early Career   
13                                       Data Modeler   
14                                     Data Scientist   
15                         Experienced Data Scientist   
16                          Data Scientist - Contract   
17                                     Data Scientist   
18                                    Data Analyst II   
19                              Medical Lab Scientist   
20                                     Data Scientist   
21                    Data Scientist/Machine Learning   
22                            Human Factors Scientist   
23     Business Intelligence Analyst I- Data Insights   
24                                     Data Scientist   
25                                     Data Scientist   
26                              Data Scientist - Risk   
27                     Data Scientist-Human Resources   
28                                     Data Scientist   
29                                     Data Scientist   
30                                     Data Scientist   
31           Data Scientist / Machine Learning Expert   
32                                     Data Scientist   
33       Senior Research Statistician- Data Scientist   
34                                      Data Engineer   
35                           Associate Data Scientist   
36                      Business Intelligence Analyst   
37                                     Data Scientist   
38                                     Data Scientist   
39                      Senior Analyst/Data Scientist   
40                                       Data Analyst   
41                          Machine Learning Engineer   
42                                     Data Analyst I   
43                      Scientist - Molecular Biology   
44                                     Data Scientist   
45          Computational Scientist, Machine Learning   
46                              Senior Data Scientist   
47                                     Data Scientist   
48                                     Data Scientist   
49                                     Data Scientist   

                    Salary Estimate  \
index                                 
0      $137K-$171K (Glassdoor est.)   
1      $137K-$171K (Glassdoor est.)   
2      $137K-$171K (Glassdoor est.)   
3      $137K-$171K (Glassdoor est.)   
4      $137K-$171K (Glassdoor est.)   
5      $137K-$171K (Glassdoor est.)   
6      $137K-$171K (Glassdoor est.)   
7      $137K-$171K (Glassdoor est.)   
8      $137K-$171K (Glassdoor est.)   
9      $137K-$171K (Glassdoor est.)   
10     $137K-$171K (Glassdoor est.)   
11     $137K-$171K (Glassdoor est.)   
12     $137K-$171K (Glassdoor est.)   
13     $137K-$171K (Glassdoor est.)   
14     $137K-$171K (Glassdoor est.)   
15     $137K-$171K (Glassdoor est.)   
16     $137K-$171K (Glassdoor est.)   
17     $137K-$171K (Glassdoor est.)   
18     $137K-$171K (Glassdoor est.)   
19     $137K-$171K (Glassdoor est.)   
20     $137K-$171K (Glassdoor est.)   
21     $137K-$171K (Glassdoor est.)   
22     $137K-$171K (Glassdoor est.)   
23     $137K-$171K (Glassdoor est.)   
24     $137K-$171K (Glassdoor est.)   
25     $137K-$171K (Glassd

### Problems of Dirty Data
* Indexing starting from 0 instead of 1

* Unnecessary annotations in the salary estimate column

* Trailing ratings in the company name column

* Some jobs not relating to data

* Most of the values in the competitors column are -1

### Cleaning and transformation to be done
1. Eliminating the (Glassdoor est.) and (Employers est.) from the 'Salary estimate' column to facilitate analysis. Also, creating a new column for minimum and maximum Salary.

2. Removing trailing ratings from the 'company name' column.

3. Creating a separate column for the initials of the 'location' column.

4. Creating a separate column for the initials of the 'Headquarters' column.

5. Modifying the 'size' column by replacing 'to' with '-' and removing 'employees'.

6. Introducing a new column to indicate the ownership status (private or public) based on the 'type of ownership' column.

7. Standardizing the 'revenue' column by replacing 'Unknown / Non-Applicable' with 'Non-Applicable', changing "billion" to "B", and "million" to "M", and using '-' instead of 'to' to follow the format '1–2B (USD)' consistently.

8. Removing the 'competitors' column if more than 75% of its values are -1.

9. Removing duplicate rows.

10. Adjust the indexing to start from 1 instead of 0.

## Checking the column name, info, no of columns, and no of rows

In [104]:
csv_dataframe.columns

Index(['Job Title', 'Salary Estimate', 'Job Description', 'Rating',
       'Company Name', 'Location', 'Headquarters', 'Size', 'Founded',
       'Type of ownership', 'Industry', 'Sector', 'Revenue', 'Competitors'],
      dtype='object')

In [105]:
# Display info
csv_dataframe.info()

<class 'pandas.core.frame.DataFrame'>
Index: 672 entries, 0 to 671
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Job Title          672 non-null    object 
 1   Salary Estimate    672 non-null    object 
 2   Job Description    672 non-null    object 
 3   Rating             672 non-null    float64
 4   Company Name       672 non-null    object 
 5   Location           672 non-null    object 
 6   Headquarters       672 non-null    object 
 7   Size               672 non-null    object 
 8   Founded            672 non-null    int64  
 9   Type of ownership  672 non-null    object 
 10  Industry           672 non-null    object 
 11  Sector             672 non-null    object 
 12  Revenue            672 non-null    object 
 13  Competitors        672 non-null    object 
dtypes: float64(1), int64(1), object(12)
memory usage: 78.8+ KB


In [106]:
# Count the number of columns
len(csv_dataframe.columns)

14

In [107]:
# Count the number of rows
len(csv_dataframe.index)

672

### Description of Dataset
The dataset had no null values, and there are 14 columns and 672 rows. Also all columns are in their correct datatypes 

### Column names and meanings
1. Job Title: The specific title or name of the data-related job position being advertised (e.g., Data Analyst, Data Scientist, Data Engineer).

2. Salary Estimate: The expected salary range for the data-related job position.

3. Job Description: A brief description of the responsibilities and requirements of the data-related job

4. Rating: The rating or feedback given to the company or job position by employees or customers

5. Company Name: The name of the company offering the data-related job position.

6. Location: The location where the data job is based or where the company is located.

7. Headquarters: The main location where the company is headquartered.

8. Size: The size of the company in terms of the number of employees

9. Founded: The year in which the company was founded

10. Type of Ownership: The ownership structure of the company (e.g., public, private, government-owned)

11. Industry: The industry in which the company operates

12. Sector: A broader category that the data-related industry belongs to (e.g., IT, healthcare services, finance).

13. Revenue: The company's total revenue or income for a given period

14. Competitors: Other companies that compete in the same data-related industry or market

### Dropping duplicate rows using the .filters chained with .drop_duplicates() 

In [108]:
# Drop duplicates rows to make each row unique
csv_dataframe=csv_dataframe.filter(items=['Job Title', 'Salary Estimate', 'Job Description', 'Rating', 'Company Name', 'Location', 'Headquarters', 'Size', 'Founded', 'Type of ownership', 'Industry', 'Sector', 'Revenue', 'Competitors']).drop_duplicates()

### Count the number of rows Remainig

In [109]:
# Count the remaining rows to check if duplicates were dropped
len(csv_dataframe.index)

659

### Cleaning 'job_title' column ...

In [110]:
# Cleaning 'job_title' column...
# Change all the 'Senior' title in Job title to Sr. to give room for consistency
csv_dataframe['Job Title']=csv_dataframe['Job Title'].str.replace('Senior', 'Sr.', regex=False)

I used the str.replace() method on the 'Job Title' column to replace all occurrences of the word 'Senior' with 'Sr.' in the job titles. 

In [111]:
csv_dataframe['Job Title'].value_counts()

Job Title
Data Scientist                                            326
Data Engineer                                              26
Sr. Data Scientist                                         21
Machine Learning Engineer                                  15
Data Analyst                                               12
                                                         ... 
Data Science Instructor                                     1
Business Data Analyst                                       1
Purification Scientist                                      1
Data Engineer, Enterprise Analytics                         1
AI/ML - Machine Learning Scientist, Siri Understanding      1
Name: count, Length: 170, dtype: int64

In [112]:
#A quick view of view of value_counts of job titles showed that some jobs are not data related, i am going to remove all of them by filtering job titles that do not have any of the keywords

In [113]:
keywords = ['data', 'analyst', 'ML/AI', 'MACHINE LEARNING', 'Machine Learning', 'Analytics']

In [114]:
# Create filter rows based on the keywords
csv_dataframe=csv_dataframe[csv_dataframe['Job Title'].str.contains('|'.join(keywords), case=False)]

I used the str.contains() method on the 'Job Title' column to filter the DataFrame based on whether the job titles contain any of the specified keywords. 

In [115]:
# Counting the number of rows remaining to check if those rows are gone
len(csv_dataframe)

611

### Cleaning the salary estimate column

In [116]:
# Cleaning 'salary_estimate' column...
# Remove the 'Glassdoor' and '(Employer estimate')' after each salary estimate
csv_dataframe ['Salary Estimate']=csv_dataframe ['Salary Estimate'].str.replace('(Glassdoor est.)', '').str.replace('(Employer est.)', '')

In [117]:
# Create a new column for minimum salary
csv_dataframe['minimum salary']=csv_dataframe['Salary Estimate'].str.replace('-.*', '', regex=True)

In [118]:
# Create a new column for maximum salary
csv_dataframe['maximum salary']=csv_dataframe['Salary Estimate'].str.split('-', n=1, expand=True)[1]

The first code removes some unnecessary annotations like "(Glassdoor est.)" and "(Employer est.)" in the 'Salary Estimate' column. To get my way around that, I used str.replace() again to remove these annotations by replacing them with empty strings, leaving only the clean salary figures. 
The second line of code also used the str.replace() method with a regular expression to remove everything after the dash "-" in each value of the "Salary Estimate" column. This extracts the minimum salary from each entry into a new column called 'minimum salary'.
I did the same for the maximum salary, but this time the code retained each value after the dash '-' but removed the ones before the dash.

### Cleaning company name column

In [119]:
# Cleaning 'company_name' column...
# Remove the \n and rating in front of the companies names
csv_dataframe['Company Name']=csv_dataframe ['Company Name'].str.replace(r"\n\d+(\.\d+)?", "", regex=True)

I used the str.replace() with the regular expression r"\n\d+(\.\d+)?" to match \n followed by a the decimal values, to replace the patterns with an empty string.

### Checking what I've done since 

In [120]:
csv_dataframe.head(5)

Job Title Salary Estimate  \
index                                      
0      Sr Data Scientist    $137K-$171K    
1         Data Scientist    $137K-$171K    
2         Data Scientist    $137K-$171K    
3         Data Scientist    $137K-$171K    
4         Data Scientist    $137K-$171K    

                                         Job Description  Rating  \
index                                                              
0      Description\n\nThe Senior Data Scientist is re...     3.1   
1      Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2      Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3      JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   
4      Data Scientist\nAffinity Solutions / Marketing...     2.9   

             Company Name       Location            Headquarters  \
index                                                              
0             Healthfirst   New York, NY            New York, NY   
1                 ManTech  Chantilly, VA             Herndon, VA   
2          Analysis Group     Boston, MA              Boston, MA   
3                 INFICON     Newton, MA  Bad Ragaz, Switzerland   
4      Affinity Solutions   New York, NY            New York, NY   

                          Size  Founded        Type of ownership  \
index                                                              
0       1001 to 5000 employees     1993   Nonprofit Organization   
1      5001 to 10000 employees     1968         Company - Public   
2       1001 to 5000 employees     1981  Private Practice / Firm   
3        501 to 1000 employees     2000         Company - Public   
4          51 to 200 employees     1998        Company - Private   

                                    Industry             Sector  \
index                                                             
0                         Insurance Carriers          Insurance   
1                     Research & Development  Business Services   
2                                 Consulting  Business Services   
3      Electrical & Electronic Manufacturing      Manufacturing   
4                    Advertising & Marketing  Business Services   

                          Revenue  \
index                               
0        Unknown / Non-Applicable   
1          $1 to $2 billion (USD)   
2      $100 to $500 million (USD)   
3      $100 to $500 million (USD)   
4        Unknown / Non-Applicable   

                                             Competitors minimum salary  \
index                                                                     
0                EmblemHealth, UnitedHealth Group, Aetna          $137K   
1                                                     -1          $137K   
2                                                     -1          $137K   
3      MKS Instruments, Pfeiffer Vacuum, Agilent Tech...          $137K   
4                   Commerce Signals, Cardlytics, Yodlee          $137K   

      maximum salary  
index                 
0             $171K   
1             $171K   
2             $171K   
3             $171K   
4             $171K

### Cleaning the Size column

In [121]:
csv_dataframe['Size']=csv_dataframe['Size'].str.replace('to', '-', regex=False)

In [122]:
# Remove the 'employees'
csv_dataframe['Size'] = csv_dataframe['Size'].str.strip('employees')

In [123]:
# Remove the space to give room for consistency
csv_dataframe['Size'] = csv_dataframe['Size'].str.replace(' ', '')

I performed 3 operations on the 'Size' column of the DataFrame 'csv_dataframe'. First, I replaced 'to' with a hyphen ('-') to transform ranges like '100 to 500 employees' into '100-500 employees'. Then, I stripped the word 'employees' from the right end of each value. Finally, I removed all spaces from the 'Size' values.

### Cleaning the rating column

In [124]:
# Cleaning 'rating' column...
# Change the -1s in the rating column to the average of all ratings
aggregate = csv_dataframe['Rating'].mean().round(1)
column_name = csv_dataframe.loc[csv_dataframe['Rating'] < 1, 'Rating'] = aggregate

I calculated the mean of the 'Rating' column in the DataFrame 'csv_dataframe' and rounded the result to one decimal place. Then, I replaced the 'Rating' values that are less than 1 with the calculated mean value.

### Cleaning the competitors column

In [125]:
# Cleaning 'competitor's' column
# Find the percentage of -1 in the competitor's column. If it's more than 40%, truncate the column
negative_rows = csv_dataframe['Competitors'].value_counts()['-1']

In [126]:
# Gettting the total number of ros
total_rows = len(csv_dataframe.index)

In [127]:
# Percentage calculation
percentage_negative_rows = (negative_rows / total_rows) * 100

In [128]:
# Round to 1 decimal place
print(percentage_negative_rows.round(1))  

75.1


In [129]:
# Total of 75.1%...Delete column

In [130]:
# Delete competitor's column
del csv_dataframe['Competitors']

I counted the occurrences of '-1' in the 'Competitors' column and calculated the total number of rows. Then, I determined the percentage of rows that contain '-1' in the 'Competitors' column. I printed the percentage with one decimal place and deleted the 'Competitors' column from the DataFrame because the occurrences were more than 70%

### Testing

In [131]:
csv_dataframe.head(5)

Job Title Salary Estimate  \
index                                      
0      Sr Data Scientist    $137K-$171K    
1         Data Scientist    $137K-$171K    
2         Data Scientist    $137K-$171K    
3         Data Scientist    $137K-$171K    
4         Data Scientist    $137K-$171K    

                                         Job Description  Rating  \
index                                                              
0      Description\n\nThe Senior Data Scientist is re...     3.1   
1      Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2      Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3      JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   
4      Data Scientist\nAffinity Solutions / Marketing...     2.9   

             Company Name       Location            Headquarters        Size  \
index                                                                          
0             Healthfirst   New York, NY            New York, NY   1001-5000   
1                 ManTech  Chantilly, VA             Herndon, VA  5001-10000   
2          Analysis Group     Boston, MA              Boston, MA   1001-5000   
3                 INFICON     Newton, MA  Bad Ragaz, Switzerland    501-1000   
4      Affinity Solutions   New York, NY            New York, NY      51-200   

       Founded        Type of ownership  \
index                                     
0         1993   Nonprofit Organization   
1         1968         Company - Public   
2         1981  Private Practice / Firm   
3         2000         Company - Public   
4         1998        Company - Private   

                                    Industry             Sector  \
index                                                             
0                         Insurance Carriers          Insurance   
1                     Research & Development  Business Services   
2                                 Consulting  Business Services   
3      Electrical & Electronic Manufacturing      Manufacturing   
4                    Advertising & Marketing  Business Services   

                          Revenue minimum salary maximum salary  
index                                                            
0        Unknown / Non-Applicable          $137K         $171K   
1          $1 to $2 billion (USD)          $137K         $171K   
2      $100 to $500 million (USD)          $137K         $171K   
3      $100 to $500 million (USD)          $137K         $171K   
4        Unknown / Non-Applicable          $137K         $171K

### Cleaning Revenue column

In [132]:
# Cleaning 'revenue' column
# Remove the 'Unknown' from revenue 'column' to remain just 'Non-Applicaple'
csv_dataframe['Revenue'] = csv_dataframe['Revenue'].str.strip('Unknown /')

In [133]:
# Replace all the '-1' with 'Non-Applicable'
csv_dataframe['Revenue'] = csv_dataframe['Revenue'].str.replace('-1', 'Non-Applicable', regex=False)

In [134]:
# Replace 'to' with '-'
csv_dataframe['Revenue'] = csv_dataframe['Revenue'].str.replace('to', '-')  

In [135]:
# Replace 'million' with 'M'
csv_dataframe['Revenue'] = csv_dataframe['Revenue'].str.replace('million', 'M')  

In [136]:
 # Replace 'billion' with 'B'
csv_dataframe['Revenue'] = csv_dataframe['Revenue'].str.replace('billion', 'B') 

In [137]:
# Strip white spaces
csv_dataframe['Revenue'] = csv_dataframe['Revenue'].str.replace(' ', '')  

In [138]:
 # Replace '+B' with 'B'
csv_dataframe['Revenue'] = csv_dataframe['Revenue'].str.replace('+B', 'B')

In [139]:
 # Replace '(USD)' with ''
csv_dataframe['Revenue'] = csv_dataframe['Revenue'].str.replace('(USD)', '') 

In [140]:
# Replace $ with ''
csv_dataframe['Revenue'] = csv_dataframe['Revenue'].str.replace('$', '') 

I performed multiple operations on the 'Revenue'. I removed the prefix 'Unknown /', replaced '-1' with 'Non-Applicable', replaced 'to' with a hyphen '-', 'million' with 'M', 'billion' with 'B', and removed spaces, '+B', '(USD)', and the dollar sign ('$'). 

### Cleaning Location column

In [141]:
# Cleaning 'Location' column...
#Split the 'Location' column into 2; one for the Location which we will
# rename 'Locationx', and the other for the location initials
split_values0 = csv_dataframe['Location'].str.split(',')

In [142]:
# Creating a new column for the location
csv_dataframe['Locationx'] = split_values0.str[0]

In [143]:
# Create another column for location initial
csv_dataframe['location_initial'] = split_values0.str[1]

In [144]:
# Delete the location column
del csv_dataframe['Location']

I splitted the 'Location' column into two separate columns. The first column, 'Locationx', contains the primary location, and the second column, 'location_initial', contains the location initials. After creating the columns, I deleted the original 'Location' column as it is no longer required.

### Testing

In [145]:
csv_dataframe.head(3)

Job Title Salary Estimate  \
index                                      
0      Sr Data Scientist    $137K-$171K    
1         Data Scientist    $137K-$171K    
2         Data Scientist    $137K-$171K    

                                         Job Description  Rating  \
index                                                              
0      Description\n\nThe Senior Data Scientist is re...     3.1   
1      Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2      Overview\n\n\nAnalysis Group is one of the lar...     3.8   

         Company Name  Headquarters        Size  Founded  \
index                                                      
0         Healthfirst  New York, NY   1001-5000     1993   
1             ManTech   Herndon, VA  5001-10000     1968   
2      Analysis Group    Boston, MA   1001-5000     1981   

             Type of ownership                Industry             Sector  \
index                                                                       
0       Nonprofit Organization      Insurance Carriers          Insurance   
1             Company - Public  Research & Development  Business Services   
2      Private Practice / Firm              Consulting  Business Services   

              Revenue minimum salary maximum salary  Locationx  \
index                                                            
0      Non-Applicable          $137K         $171K    New York   
1                1-2B          $137K         $171K   Chantilly   
2            100-500M          $137K         $171K      Boston   

      location_initial  
index                   
0                   NY  
1                   VA  
2                   MA

### Cleaning the Headquarters column

In [146]:
# Cleaning 'Headquarters' column...
#Split the 'Headquarters' column into 2; one for the Headquarters which we will
# rename 'head_office', and the other for the city initials
split_values = csv_dataframe['Headquarters'].str.split(',')

In [147]:
# Creating a new column for the Headquarters and name it head_office
csv_dataframe['head_office'] = split_values.str[0]

In [148]:
# Creating a new column for the city_initials
csv_dataframe['headoffice_initials'] = split_values.str[1]

In [149]:
# Delete the headquarters column
del csv_dataframe['Headquarters']

I splitted the 'Headquarters' column into two separate columns. The first column, 'Head office', contains the primary headquarter, and the second column, 'headquarter_initial', contains the headquarters initials. After creating the columns, I deleted the original 'Headquarters' column as it is no longer required.

In [150]:
# Check if all the cities are in initials format
print(csv_dataframe['headoffice_initials'].value_counts())  

headoffice_initials
 CA                126
 VA                 79
 MA                 47
 NY                 38
 IL                 32
 NJ                 30
 MD                 28
 CO                 21
 WA                 14
 PA                 13
 United Kingdom     13
 TX                 12
 FL                 12
 OH                 11
 MO                 10
 Switzerland         9
 MN                  8
 NC                  7
 CT                  7
 TN                  7
 DC                  7
 MI                  4
 IA                  4
 IN                  4
 Canada              4
 France              4
 GA                  4
 WI                  4
 NE                  4
 UT                  3
 AZ                  3
 SC                  3
 India               3
 OR                  2
 061                 2
 Bermuda             2
 Singapore           2
 Japan               2
 AL                  2
 WV                  2
 HI                  1
 LA                  1
 Sweden       

In [151]:
# Some headquarters are not in their initials format so, i am going to map them to the correct initials using a dictionary

In [152]:
# 10 cities are not in their initials format, so we replace the cities with their initials using a dictionary
inconsistent_cities = {'headoffice_initials': ['United Kingdom', 'Switzerland', 'France', 'Canada', 'India', 'Japan',
                                         'Bermuda', 'Singapore', 'Spain', 'Sweden', 'Israel']}

In [153]:
new_initials = {'United Kingdom': 'UK', 'Switzerland': 'CH', 'France': 'FR', 'Canada': 'CA',
                'India': 'IN', 'Japan': 'JP', 'Bermuda': 'BERM', 'Singapore': 'SG', 'Spain': 'ES',
                'Sweden': 'SE', 'Israel': 'IL'}

In [154]:
csv_dataframe['headoffice_initials'] = csv_dataframe['headoffice_initials'].replace(new_initials, regex=True)

In [155]:
csv_dataframe.head(4)

Job Title Salary Estimate  \
index                                      
0      Sr Data Scientist    $137K-$171K    
1         Data Scientist    $137K-$171K    
2         Data Scientist    $137K-$171K    
3         Data Scientist    $137K-$171K    

                                         Job Description  Rating  \
index                                                              
0      Description\n\nThe Senior Data Scientist is re...     3.1   
1      Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2      Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3      JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   

         Company Name        Size  Founded        Type of ownership  \
index                                                                 
0         Healthfirst   1001-5000     1993   Nonprofit Organization   
1             ManTech  5001-10000     1968         Company - Public   
2      Analysis Group   1001-5000     1981  Private Practice / Firm   
3             INFICON    501-1000     2000         Company - Public   

                                    Industry             Sector  \
index                                                             
0                         Insurance Carriers          Insurance   
1                     Research & Development  Business Services   
2                                 Consulting  Business Services   
3      Electrical & Electronic Manufacturing      Manufacturing   

              Revenue minimum salary maximum salary  Locationx  \
index                                                            
0      Non-Applicable          $137K         $171K    New York   
1                1-2B          $137K         $171K   Chantilly   
2            100-500M          $137K         $171K      Boston   
3            100-500M          $137K         $171K      Newton   

      location_initial head_office headoffice_initials  
index                                                   
0                   NY    New York                  NY  
1                   VA     Herndon                  VA  
2                   MA      Boston                  MA  
3                   MA   Bad Ragaz                  CH

### Cleaning type of ownership column

In [156]:
# Cleaning 'type_of_ownership' column...
# Replace 'Government' with 'Government-Public' because we will be splitting it and renaming
# ,and we don't want Nan in the status of ownership

In [157]:
# Replacing 'Government' with 'Government-Public'
csv_dataframe['Type of ownership'] = csv_dataframe['Type of ownership'].str.replace('Government', 'Government-Public')

In [158]:
# Replacing 'College/University' with 'College' for consistency
csv_dataframe['Type of ownership'] = csv_dataframe['Type of ownership'].str.replace('College / University', 'College')

In [159]:
# Splitting the column after the occurrence of '-'
split_values = csv_dataframe['Type of ownership'].str.split('-')

In [160]:
# Naming the words before the occurrence of '-' 'ownership'
csv_dataframe['ownership'] = split_values.str[0]

In [161]:
# Naming the words after the occurrence of '-' 'status'
csv_dataframe['status'] = split_values.str[1]

In [162]:
# Fill the NaN in the status column with Not Specified
csv_dataframe['status'].fillna('Not Specified', inplace=True)

In [163]:
 # Dropping the original column
del csv_dataframe['Type of ownership']

I modified the 'Type of ownership' column by replacing 'Government' with 'Government-Public' and 'College / University' with 'College'. Then, I split the 'Type of ownership' values by the hyphen separator and created two new columns, 'ownership' and 'status', to store the values of the 'split values'. I filled any missing values in the 'status' column with 'Not Specified'so they won't be null. Finally, I deleted the original 'Type of ownership' column as it is no longer needed.

### Still on the new column I just created 

In [164]:
# Replace all inconsistencies in ownership column
inconsistent_ownership = {'ownership': ['College / University', 'Subsidiary or Business Segment',
                                        'Nonprofit Organization', 'Other Organization']}

In [165]:
# Map inconsistencies with correct words
new_initial = {'College / University': 'College', 'Subsidiary or Business Segment': 'Subsidiary/Segment',
               'Nonprofit Organization': 'Nonprofit', 'Other Organization': 'Others'}

In [166]:
# Now replace them in the ownership column 
csv_dataframe['ownership'] = csv_dataframe['ownership'].replace(new_initial, regex=True)

I created a new column 'new_initial' as a dictionary to make the ownership types more consistent in the 'ownership' column . I replaced the inconsistent ownership types I defined in the 'inconsistent_ownership' dictionary with their consistent representations.

In [167]:
# Now let's check the dataframe so far
csv_dataframe.head(5) 

Job Title Salary Estimate  \
index                                      
0      Sr Data Scientist    $137K-$171K    
1         Data Scientist    $137K-$171K    
2         Data Scientist    $137K-$171K    
3         Data Scientist    $137K-$171K    
4         Data Scientist    $137K-$171K    

                                         Job Description  Rating  \
index                                                              
0      Description\n\nThe Senior Data Scientist is re...     3.1   
1      Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2      Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3      JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   
4      Data Scientist\nAffinity Solutions / Marketing...     2.9   

             Company Name        Size  Founded  \
index                                            
0             Healthfirst   1001-5000     1993   
1                 ManTech  5001-10000     1968   
2          Analysis Group   1001-5000     1981   
3                 INFICON    501-1000     2000   
4      Affinity Solutions      51-200     1998   

                                    Industry             Sector  \
index                                                             
0                         Insurance Carriers          Insurance   
1                     Research & Development  Business Services   
2                                 Consulting  Business Services   
3      Electrical & Electronic Manufacturing      Manufacturing   
4                    Advertising & Marketing  Business Services   

              Revenue minimum salary maximum salary  Locationx  \
index                                                            
0      Non-Applicable          $137K         $171K    New York   
1                1-2B          $137K         $171K   Chantilly   
2            100-500M          $137K         $171K      Boston   
3            100-500M          $137K         $171K      Newton   
4      Non-Applicable          $137K         $171K    New York   

      location_initial head_office headoffice_initials  \
index                                                    
0                   NY    New York                  NY   
1                   VA     Herndon                  VA   
2                   MA      Boston                  MA   
3                   MA   Bad Ragaz                  CH   
4                   NY    New York                  NY   

                     ownership         status  
index                                          
0                    Nonprofit  Not Specified  
1                     Company          Public  
2      Private Practice / Firm  Not Specified  
3                     Company          Public  
4                     Company         Private

### Cleaning the job title column



In [168]:
csv_dataframe['Job Title'].value_counts()



Job Title
Data Scientist                                            326
Data Engineer                                              26
Sr. Data Scientist                                         21
Machine Learning Engineer                                  15
Data Analyst                                               12
                                                         ... 
Data Engineer (Remote)                                      1
Business Data Analyst                                       1
Production Engineer - Statistics/Data Analysis              1
Data Engineer, Enterprise Analytics                         1
AI/ML - Machine Learning Scientist, Siri Understanding      1
Name: count, Length: 135, dtype: int64

I noticed some job titles have descriptions in them, so i will be mapping each job description to the correct one

In [169]:
data = {'Job Title':[
"Data Scientist",
"Data Engineer",
"Sr. Data Scientist",
"Machine Learning Engineer",
"Data Analyst",
"Sr. Data Analyst",
"Data Science Software Engineer",
"Sr. Data Engineer",
"Data Scientist - TS/SCI FSP or CI Required",
"Analytics - Business Assurance Data Analyst",
"AI Ops Data Scientist",
"Sr. Business Intelligence Analyst",
"Data Modeler (Analytical Systems)",
"Sr. Machine Learning Scientist - Bay Area, CA",
"Data Scientist - TS/SCI Required",
"Sr. Data Scientist – Image Analytics, Novartis AI Innovation Lab",
"Principal Data Scientist",
"Sr. ML/Data Scientist - AI/NLP/Chatbot",
"Lead Data Scientist",
"Health Plan Data Analyst, Sr",
"Sr. Data Scientist - R&D Oncology",
"Principal Machine Learning Scientist",
"In-Line Inspection Data Analyst",
"Principal Data Scientist - Machine Learning",
"Sr Data Scientist",
"Intelligence Data Analyst, Sr",
"(Sr.) Data Scientist -",
"Analytics Manager - Data Mart",
"Machine Learning Engineer/Scientist",
"Data Analyst - Unilever Prestige",
"VP, Data Science",
"Data Engineer - Kafka",
"Scientist - Machine Learning",
"Data Engineer (Analytics, SQL, Python, AWS)",
"Sr. Data Analyst - Finance & Platform Analytics",
"IT Partner Digital Health Technology and Data Science",
"Data Scientist (TS/SCI w/ Poly)",
"Cloud Data Engineer (Azure)",
"Sr. Machine Learning Engineer",
"Data Scientist(s)/Machine Learning Engineer",
"Data Scientist, Kinship - NYC/Portland",
"Sr. Data Scientist II",
"Patient Safety- Associate Data Scientist",
"Sr. Analyst/Data Scientist",
"Data Scientist, Applied Machine Learning - Bay Area",
"Experienced Data Scientist",
"Data Scientist - Intermediate",
"Data Scientist - Statistics, Early Career",
"Staff Data Scientist - Analytics",
"Product Data Scientist - Ads Data Science",
"Data Analytics Engineer",
"Data Scientist / Machine Learning Expert",
"Data Scientist 3 (718)",
"Data Scientist - Contract",
"Data Scientist/Machine Learning",
"Data Scientist - Machine Learning",
"Data & Machine Learning Scientist",
"Business Intelligence Analyst",
"Real World Science, Data Scientist",
"Data Scientist - Image and Video Analytics",
"Associate Data Scientist",
"Data Scientist - Risk",
"Data Scientist / Applied Mathematician",
"Data Scientist-Human Resources",
"Sr. Research Statistician- Data Scientist",
"Data Analyst II",
"Business Intelligence Analyst I- Data Insights",
"Principal Data & Analytics Platform Engineer",
"Market Research Data Scientist",
"Data Scientist Machine Learning",
"Diversity and Inclusion Data Analyst",
"Data Science All Star Program - Data Engineer Track",
"Data Solutions Engineer - Data Modeler",
"Software Engineer (Data Scientist, C,C++,Linux,Unix) - SISW - MG",
"Sr. Clinical Data Scientist Programmer",
"Data Scientist/Data Analytics Practitioner",
"Developer III - Data Science",
"Lead Data Scientist – Network Analysis and Control",
"Enterprise Data Analyst (Enterprise Portfolio Management Office)",
"Vice President, Biometrics and Clinical Data Management",
"Machine Learning Engineer, Sr.",
"Aviation AI/ML Data Scientist",
"Data Architect",
"Big Data Engineer",
"Data Science Manager",
"Data Scientist Technical Specialist",
"Clinical Data Analyst",
"Software Engineer - Machine Learning & Data Science (Applied Intelligence Services Team)",
"Data Science Analyst",
"Machine Learning Scientist / Engineer",
"Sr. Principal Data Scientist (Python/R)",
"Data Scientist- Industrial Discrete Sector Industry",
"Data Modeler",
"Data Engineer, Digital & Comp Pathology",
"Manager / Lead, Data Science & Analytics",
"Equity Data Insights Analyst - Quantitative Analyst",
"Director of Data Science",
"Data Analyst/Engineer",
"Sr Data Analyst",
"Geospatial Data Scientist",
"Say Business Data Analyst",
"Analytics Manager",
"AI Data Scientist",
"Tableau Data Engineer",
"Data Integration and Modeling Engineer",
"Software Engineer - Data Science",
"Sr. Data Scientist - Algorithms",
"Data Scientist (TS/SCI)",
"RFP Data Analyst",
"Data Science Manager, Payment Acceptance - USA",
"Sr Data Engineer (Sr BI Developer)",
"Staff Data Scientist",
"Health Data Scientist - Biomedical/Biostats",
"Applied Technology Researcher / Data Scientist",
"Machine Learning Scientist - Bay Area, CA",
"Sr. Data & Machine Learning Scientist",
"Data Scientist - Statistics, Mid-Career",
"Computational Scientist, Machine Learning",
"Staff BI and Data Engineer",
"Environmental Data Science",
"Software Data Engineer",
"Operations Data Analyst",
"Staff Data Scientist - Pricing",
"Report Writer-Data Analyst",
"Data Analyst I",
"Jr. Data Engineer",
"Jr. Business Data Analyst (position added)",
"E-Commerce Data Analyst",
"Data Science Instructor",
"Global Data Analyst",
"Data Engineer (Remote)",
"Business Data Analyst",
"Production Engineer - Statistics/Data Analysis",
"Data Engineer, Enterprise Analytics",
"AI/ML - Machine Learning Scientist, Siri Understanding",
"Sr. Data Scientist – Image Analytics, Novartis AI Innovation Lab",
"AI/ML - Machine Learning Scientist, Siri Understanding"]}

In [170]:
df = pd.DataFrame(data)

In [171]:
new_job_titles={
'Data Scientist' :'Data Scientist',
'Data Engineer' :'Data Engineer',
'Sr. Data Scientist' :'Sr. Data Scientist',
'Machine Learning Engineer':  'Machine Learning Engineer',
'Data Analyst':'Data Analyst',
'Sr. Data Analyst':'Sr. Data Analyst',
'Sr. Data Engineer': 'Sr. Data Engineer',
'Data Science Software Engineer':'Data Science Software Engineer',
'Analytics Manager': 'Analytics Manager',
'AI Ops Data Scientist': 'AI Ops Data Scientist',
'Principal Data Scientist' : 'Principal Data Scientist',
'Sr. ML/Data Scientist' :'Sr. ML/Data Scientist',
'Data Modeler (Analytical Systems)':'Data Modeler',
'Sr. Business Intelligence Analyst': 'Sr. Business Intelligence Analyst',
'Business Assurance Data Analyst': 'Business Assurance Data Analyst',
'Lead Data Scientist': 'Lead Data Scientist',
'Sr. Machine Learning Scientist':'Sr. Machine Learning Scientist',
'Data Scientist(s)/Machine Learning Engineer': 'Data Scientist(s)/Machine Learning Engineer',
'In-Line Inspection Data Analyst':  'In-Line Inspection Data Analyst',
'Data Scientist / Applied Mathematician':  'Data Scientist / Applied Mathematician',
'Patient Safety- Associate Data Scientist':  'Associate Data Scientist',
'(Sr.) Data Scientist -': '(Sr.) Data Scientist',
'VP, Data Science':'VP, Data Science',
'Scientist - Machine Learning': 'Machine Learning Scientist',
'Data Engineer - Kafka': 'Data Engineer',
'Sr. Data Scientist II':'Sr. Data Scientist II',
'Data Engineer (Analytics, SQL, Python, AWS)': 'Data Engineer',
'Real World Science, Data Scientist': 'Data Scientist',
'Sr. Machine Learning Engineer': 'Sr. Machine Learning Engineer',
'Cloud Data Engineer (Azure)': 'Cloud Data Engineer',
'Data Scientist (TS/SCI w/ Poly)': 'Data Scientist',
'Sr. Data Analyst - Finance & Platform Analytics': 'Sr. Data Analyst',
'IT Partner Digital Health Technology and Data Science': 'IT Partner and Data Scientist',
'Principal Machine Learning Scientist':'Principal Machine Learning Scientist',
'Principal Data Scientist - Machine Learning': 'Principal Data Scientist - Machine Learning',
'Health Plan Data Analyst, Sr':'Sr. Health Plan Data Analyst',
'Sr Data Scientist': 'Sr Data Scientist',
'Data Scientist 3 (718)':'Data Scientist',
'Business Intelligence Analyst':'Business Intelligence Analyst',
'Data Scientist / Machine Learning Expert':'Data Scientist / Machine Learning Expert',
'Staff Data Scientist - Analytics': 'Staff Data Scientist',
'Data Scientist - Statistics, Early Career': 'Data Scientist',
'Experienced Data Scientist': 'Data Scientist',
'Data Scientist - Contract': 'Data Scientist',
'Data Scientist/Machine Learning': 'Data Scientist/Machine Learning',
'Data Scientist - Risk': 'Data Scientist',
'Data Scientist-Human Resources': 'Human Resources Data Scientist',
'Sr. Research Statistician- Data Scientist':'Sr. Research Statistician- Data Scientist',
'Associate Data Scientist':'Associate Data Scientist',
'Sr. Analyst/Data Scientist':'Sr. Analyst/Data Scientist',
'Machine Learning Engineer/Scientist': 'Machine Learning Engineer/Scientist',
'Data Analytics Engineer':'Data Analytics Engineer',
'Product Data Scientist':'Product Data Scientist',
'Data Scientist - Intermediate':'Data Scientist',
'Data & Machine Learning Scientist': 'Data & Machine Learning Scientist',
'Data Scientist - Machine Learning':'Machine Learning Data Scientist',
'Intelligence Data Analyst, Sr.':'Sr. Intelligence Data Analyst',
'Data Scientist, Applied Machine Learning - Bay Area':'Data Scientist, Applied Machine Learning',
'Data Science All Star Program - Data Engineer Track':'Data Science Program - Data Engineer Track',
'Data Scientist Machine Learning':'Machine Learning Data Scientist',
'Principal Data & Analytics Platform Engineer': 'Principal Data & Analytics Platform Engineer',
'Diversity and Inclusion Data Analyst':'Diversity and Inclusion Data Analyst',
'Manager / Lead, Data Science & Analytics': 'Lead, Data Science & Analytics',
'Data Engineer, Digital & Comp Pathology':'Data Engineer',
'Data Science Analyst':'Data Science Analyst',
'Market Research Data Scientist':'Market Research Data Scientist',
'Big Data Engineer':'Big Data Engineer',
'Data Scientist/Data Analytics Practitioner': 'Data Scientist/Data Analytics Practitioner',
'Developer III - Data Science':'Data Scientist',
'Lead Data Scientist – Network Analysis and Control':'Lead Data Scientist',
'Enterprise Data Analyst (Enterprise Portfolio Management Office)':'Enterprise Data Analyst',
'Vice President, Biometrics and Clinical Data Management':'Vice President, Biometrics and Clinical Data Management',
'Machine Learning Engineer, Sr.':'Sr.Machine Learning Engineer',
'Aviation AI/ML Data Scientist': 'Aviation AI/ML Data Scientist',
'Data Architect': 'Data Architect',
'Data Science Manager':'Data Science Manager',
'Software Engineer (Data Scientist, C,C++,Linux,Unix) - SISW - MG':'Software Engineer (Data Scientist)',
'Data Scientist Technical Specialist':'Data Scientist Technical Specialist',
'Clinical Data Analyst':'Clinical Data Analyst',
'Software Engineer - Machine Learning & Data Science (Applied Intelligence Services Team)': 'Software Engineer - Machine Learning & Data Science',
'Data Scientist - Statistics, Mid-Career':'Data Scientist - Statistics',
'Machine Learning Scientist / Engineer':'Machine Learning Scientist / Engineer',
'Sr. Principal Data Scientist (Python/R)': 'Sr. Principal Data Scientist',
'Data Solutions Engineer - Data Modeler':'Data Solutions Engineer - Data Modeler',
'Sr. Clinical Data Scientist Programmer':'Sr. Clinical Data Scientist Programmer',
'Director of Data Science': 'Director of Data Science',
'Sr. Data & Machine Learning Scientist':'Sr. Data & Machine Learning Scientist',
'Staff BI and Data Engineer': 'Staff BI and Data Engineer',
'Data Engineer (Remote)': 'Data Engineer',
'Sr Data Engineer (Sr BI Developer)': 'Sr Data Engineer (Sr BI Developer)',
'Staff Data Scientist':  'Staff Data Scientist',
'Health Data Scientist - Biomedical/Biostats': 'Health Data Scientist',
'Applied Technology Researcher / Data Scientist':'Applied Technology Researcher / Data Scientist',
'Data Science Manager, Payment Acceptance - USA':'Data Science Manager',
'Data Engineer, Enterprise Analytics':'Data Engineer',
'Business Data Analyst':'Business Data Analyst',
'Global Data Analyst':'Global Data Analyst',
'Data Scientist (TS/SCI)':'Data Scientist',
'E-Commerce Data Analyst':'E-Commerce Data Analyst',
'Jr. Data Engineer': 'Jr. Data Engineer',
'Computational Scientist, Machine Learning':'Computational Scientist, Machine Learning',
'Data Analyst I':'Data Analyst I',
'Business Intelligence Analyst I- Data Insights':'Business Intelligence Analyst I',
'Data Analyst II':'Data Analyst II',
'Data Modeler':'Data Modeler',
'RFP Data Analyst': 'RFP Data Analyst',
'Software Engineer - Data Science':'Software Engineer - Data Science',
'Environmental Data Science':'Environmental Data Science',
'Jr. Business Data Analyst (position added 6/12/2020)':'Jr. Business Data Analyst',
'Software Data Engineer':'Software Data Engineer',
'Operations Data Analyst':'Operations Data Analyst',
'Equity Data Insights Analyst - Quantitative Analyst':'Equity Data Insights Analyst - Quantitative Analyst',
'Staff Data Scientist - Pricing':'Staff Data Scientist',
'Report Writer-Data Analyst':'Report Writer-Data Analyst',
'Data Science Instructor': 'Data Science Instructor',
'Production Engineer - Statistics/Data Analysis': 'Production Engineer - Statistics/Data Analysis',
'Sr. Data Scientist - Algorithms':'Sr. Data Scientist',
'Data Analyst/Engineer':'Data Analyst/Engineer',
'Machine Learning Scientist - Bay Area, CA' :'Machine Learning Scientist',
'Sr Data Analyst': 'Sr Data Analyst',
'Geospatial Data Scientist': 'Geospatial Data Scientist',
'Say Business Data Analyst': 'Business Data Analyst',
'AI Data Scientist':'AI Data Scientist',
'Tableau Data Engineer 20-0117': 'Tableau Data Engineer',
'Data Integration and Modeling Engineer':'Data Integration and Modeling Engineer',
'AI/ML - Machine Learning Scientist':'AI/ML - Machine Learning Scientist',
'Sr. Data Scientist – Image Analytics, Novartis AI Innovation Lab':'Sr. Data Scientist',
'AI/ML - Machine Learning Scientist, Siri Understanding':'AI/ML - Machine Learning Scientist'}

In [172]:
csv_dataframe['Job Title']=csv_dataframe['Job Title'].replace(new_job_titles)

I created a Dataframe consisting of all job titles in the value counts then mapped it for consistency then i replaced the job titles with the new ones i just created

In [173]:
csv_dataframe ['Job Title'] = csv_dataframe ['Job Title'].apply(lambda x: x[:7] + x[7:].split('-')[0])

Removing any character after the seventh character of each job title. This removes any text after the first '-' character

In [175]:
csv_dataframe.head(5)

Job Title Salary Estimate  \
index                                      
0      Sr Data Scientist    $137K-$171K    
1         Data Scientist    $137K-$171K    
2         Data Scientist    $137K-$171K    
3         Data Scientist    $137K-$171K    
4         Data Scientist    $137K-$171K    

                                         Job Description  Rating  \
index                                                              
0      Description\n\nThe Senior Data Scientist is re...     3.1   
1      Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2      Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3      JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   
4      Data Scientist\nAffinity Solutions / Marketing...     2.9   

             Company Name        Size  Founded  \
index                                            
0             Healthfirst   1001-5000     1993   
1                 ManTech  5001-10000     1968   
2          Analysis Group   1001-5000     1981   
3                 INFICON    501-1000     2000   
4      Affinity Solutions      51-200     1998   

                                    Industry             Sector  \
index                                                             
0                         Insurance Carriers          Insurance   
1                     Research & Development  Business Services   
2                                 Consulting  Business Services   
3      Electrical & Electronic Manufacturing      Manufacturing   
4                    Advertising & Marketing  Business Services   

              Revenue minimum salary maximum salary  Locationx  \
index                                                            
0      Non-Applicable          $137K         $171K    New York   
1                1-2B          $137K         $171K   Chantilly   
2            100-500M          $137K         $171K      Boston   
3            100-500M          $137K         $171K      Newton   
4      Non-Applicable          $137K         $171K    New York   

      location_initial head_office headoffice_initials  \
index                                                    
0                   NY    New York                  NY   
1                   VA     Herndon                  VA   
2                   MA      Boston                  MA   
3                   MA   Bad Ragaz                  CH   
4                   NY    New York                  NY   

                     ownership         status  
index                                          
0                    Nonprofit  Not Specified  
1                     Company          Public  
2      Private Practice / Firm  Not Specified  
3                     Company          Public  
4                     Company         Private

In [176]:
# Delete the index column since python automatically index all rows and reset it to start from 1 not 0
csv_dataframe = csv_dataframe.reset_index(drop=True)

In [177]:
# Set the index to start from 1 not 0
csv_dataframe.index = csv_dataframe.index + 1

In [178]:
# Save cleaned data to Excel file
csv_dataframe.to_excel(r'C:\Users\idowu\Documents\datasciencejobs.xlsx', index=False)